### ISSUE TITLE
Inheritance :: error when adding Customer and VipCustomer to a db set in one "save changes"

### ISSUE THREAD

**Issue by maumar** [aspnet/EntityFramework#3161]
Repro:

```
using System.Data.SqlClient;
using Microsoft.Data.Entity;

namespace Repro3161
{
    public class Customer
    {
        public int Id { get; set; }
        public string Name { get; set; }
    }

    public class VipCustomer : Customer
    {
    }

    public class MyContext : DbContext
    {
        public DbSet<Customer> Customers { get; set; }

        protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
        {
            optionsBuilder.UseSqlServer(new SqlConnectionStringBuilder { DataSource = ".", InitialCatalog = "Repro3161", MultipleActiveResultSets = true, IntegratedSecurity = true }.ToString());
        }

        protected override void OnModelCreating(ModelBuilder modelBuilder)
        {
            modelBuilder.Entity<VipCustomer>().BaseType<Customer>();
        }
    }

    class Program
    {
        static void Main(string[] args)
        {
            using (var ctx = new MyContext())
            {
                ctx.Database.EnsureDeleted();
                ctx.Database.EnsureCreated();

                var c1 = new Customer { Name = "C1", };
                var c2 = new Customer { Name = "C2", };
                var c3 = new VipCustomer { Name = "VC3", };

                ctx.Customers.AddRange(c1, c2, c3);
                ctx.SaveChanges();
            }
        }
    }
}

```

This throws:
The instance of entity type 'Repro3161.VipCustomer' cannot be tracked because another instance of this type with the same key is already being tracked. For new entities consider using an IIdentityGenerator to generate unique key values.

   at Microsoft.Data.Entity.ChangeTracking.Internal.StateManager.StartTracking(InternalEntityEntry entry) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\StateManager.cs:line 222
   at Microsoft.Data.Entity.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState oldState, EntityState newState, Boolean acceptChanges) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\InternalEntityEntry.cs:line 177
   at Microsoft.Data.Entity.ChangeTracking.Internal.InternalEntityEntry.SetEntityState(EntityState entityState, Boolean acceptChanges) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\InternalEntityEntry.cs:line 101
   at Microsoft.Data.Entity.ChangeTracking.Internal.EntityGraphAttacher.PaintAction(InternalEntityEntryGraphNode n) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\EntityGraphAttacher.cs:line 39
   at Microsoft.Data.Entity.ChangeTracking.Internal.EntityEntryGraphIterator.TraverseGraph[TNode](TNode node, Func`2 handleNode) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\EntityEntryGraphIterator.cs:line 15
   at Microsoft.Data.Entity.ChangeTracking.Internal.EntityGraphAttacher.AttachGraph(InternalEntityEntry rootEntry, EntityState entityState) in D:\k\EntityFramework\src\EntityFramework.Core\ChangeTracking\Internal\EntityGraphAttacher.cs:line 19
   at Microsoft.Data.Entity.DbContext.SetEntityState(InternalEntityEntry entry, EntityState entityState, GraphBehavior behavior) in D:\k\EntityFramework\src\EntityFramework.Core\DbContext.cs:line 449
   at Microsoft.Data.Entity.DbContext.SetEntityStates(IEnumerable`1 entities, EntityState entityState, GraphBehavior behavior) in D:\k\EntityFramework\src\EntityFramework.Core\DbContext.cs:line 717
   at Microsoft.Data.Entity.DbContext.AddRange(IEnumerable`1 entities, GraphBehavior behavior) in D:\k\EntityFramework\src\EntityFramework.Core\DbContext.cs:line 732
   at Microsoft.Data.Entity.DbContext.AddRange(Object[] entities) in D:\k\EntityFramework\src\EntityFramework.Core\DbContext.cs:line 673
   at Microsoft.Data.Entity.Internal.InternalDbSet`1.AddRange(TEntity[] entities) in D:\k\EntityFramework\src\EntityFramework.Core\Internal\InternalDbSet.cs:line 59

**Comment by maumar** [comment]
Everything works fine if VipCustomer is added in separate SaveChanges

**Comment by smitpatel** [comment]
The CacheKey for ValueGeneratorCache is defined using combination of Property & EntityType. Since for Base & Derived class both have different EntityType, both uses different value generator therefore values for PK clashes and exception is thrown.

**Comment by AndriySvyryd** [comment]
This should be fixed by https://github.com/aspnet/EntityFramework/pull/3441

**Comment by smitpatel** [comment]
fixed by 369ea75487e57bbe6b80528d2a246d6f83b1f6b1
